README.md

```diff
@@ -221,26 +221,44 @@ generally advisable to allow the keeper to manage gas prices for bids, and not s
-from the `Vat`.
-
-**Using the `eth-from` account on multiple keepers is also discouraged** as it complicates `Vat` inventory management.
-When running multiple keepers using the same account, the balance of Dai in the `Vat` will be shared across keepers.  
-If using the feature, set `--vat-dai-target` to the same value on each keeper, and sufficiently high to cover total
-desired exposure.
-
-To manually control the amount of Dai in the `Vat`, pass `--keep-dai-in-vat-on-exit` and `--keep-gem-in-vat-on-exit`
-switches, and do not pass the `--vat-dai-target` switch.  You may use [mcd-cli](https://github.com/makerdao/mcd-cli)
-to manually `join`/`exit` Dai to/from each of your keeper accounts.  Here is an example to join 6000 Dai on a testnet,
+Key points:
+- Dai must be **joined** from a token balance to the `Vat` for bidding on `flip` and `flop` auctions.
+- Won collateral can be **exited** from the `Vat` to a token balance after a won auction is dealt (closed).
+- MKR for/from `flap`/`flop` auctions is managed directly through token balances and is never joined to the `Vat`.
+
+The keeper provides facilities for managing `Vat` balances, which may be turned off to manage manually. 
+To manually control the amount of Dai in the `Vat`, pass `--keep-dai-in-vat-on-exit`, set `--return-gem-behavior NONE`, 
+and do not pass the `--vat-dai-target` switch.
+
+Warnings: **Do not use an `eth-from` account on multiple keepers** as it complicates Vat inventory management and 
+will likely cause nonce conflicts.  Using an `eth-from` account with an open vault is also discouraged.
+
+#### Dai
+All auction contracts exclusively interact with Dai (for all auctions) in the `Vat`.  `--vat-dai-target` may be set to 
+the amount you wish to maintain, or `all` to join your account's entire token balance.  Rebalances do not account for 
+Dai moved from the `Vat` to an auction contract for an active bid.  There are three places where Dai is or can be 
+rebalanced, all using `--vat-dai-target`:
+- Upon keeper startup
+- Periodically, if `--rebalance-interval` is set (see _Rebalancing_ section below)
+- When when an auction is dealt
+
+By default, all Dai in your `eth-from` account is exited from the `Vat` and added to your token balance when the keeper 
+is shut down.  This feature may be disabled using `--keep-dai-in-vat-on-exit`.
+
+#### Collateral (flip auctions)
+`--return-gem-behavior` offers a few options for exiting won collateral to a token balance:
+- `ONEXIT` - (default) collateral is returned when the keeper exits normally
+- `ONREBALANCE` - collateral balances are withdrawn upon periodic rebalances (discussed below)
+- `NONE` - the keeper doesn't return collateral (user is responsible for withdrawing from `Vat` manually)
+
+#### Rebalancing
+Dai can be periodically joined or exited  by setting `--rebalance-interval` to the number of seconds between balance 
```

Comment:
can we consider joining DAI only on demand? e.g. keeper wants to bid, it calculates DAI needed and figures out there's not enough to bid, then check if DAI available to join and  if true join DAI for placing bid. This reduce configuration complexity and minimize number of txes to be send for rebalancing.

Reply:
Joining on demand would require a proxy contract to reduce the number of TXes and save gas fees.  Without a proxy contract, every bid would be delayed while awaiting a `join` to complete.  Joining upon each bid would also burn significantly more gas, as more `join` operations would need to be performed.

Reply:
then do only one tx and join all DAI when a bid with insufficient DAI

Reply:
I considered replacing periodic rebalancing with rebalancing Dai only upon starvation, but this would:
- still delay one or more bids while waiting for the join
- not help exit won collateral (which would waste gas if done after every `deal`)

This suggests a need for a behavior parameter analogous to `--return-gem-behavior`.  Here are proposed options for a `--supply-dai-behavior` switch:
- `ONSTARTUP` - (default) Dai is only joined when the keeper starts up
- `ONBID` - Dai is synchronously joined when the model attempts to bid and there is insufficient Dai to satisfy the bid.
- `ONREBALANCE` - Dai is joined upon periodic rebalances (see `--rebalance-interval`)
- `NONE` - Dai is not joined or exited by the keeper
All options honor `--vat-dai-target` regarding the amount of Dai to join; bids will not be submitted if the target is set lower than Dai required by the model's output, regardless of token balance.

This doesn't resolve awkwardness `exit`ing Dai.  We could retain the default behavior of exiting on keeper shutdown and upon rebalance (unless target is set to `all`.

Reply:
> I considered replacing periodic rebalancing with rebalancing Dai only upon starvation, but this would:
> 
> * still delay one or more bids while waiting for the join

This is valid also in case of periodic rebalancing (which can actually have longer delays than rebalancing on demand)

> * not help exit won collateral (which would waste gas if done after every `deal`)

Correct, my comment was only for DAI join

> 
> This suggests a need for a behavior parameter analogous to `--return-gem-behavior`. Here are proposed options for a `--supply-dai-behavior` switch:
> 
> * `ONSTARTUP` - (default) Dai is only joined when the keeper starts up
> * `ONBID` - Dai is synchronously joined when the model attempts to bid and there is insufficient Dai to satisfy the bid.
> * `ONREBALANCE` - Dai is joined or exited upon periodic rebalances (see `--rebalance-interval`)
> * `NONE` - Dai is not joined or exited by the keeper
>   All options honor `--vat-dai-target` regarding the amount of Dai to join; bids will not be submitted if the target is set lower than Dai required by the model's output, regardless of token balance.
> 

I find this too complicated and potentially confusing, we should really keep simple and join on demand only, so I suggest leaving join only on bid (default) and none. For exit we can again have on exit and none

> This doesn't resolve awkwardness `exit`ing Dai. We could retain the default behavior of exiting on keeper shutdown and upon rebalance (unless target is set to `all`.

Correct, my comment was only for joining, I don't feel like we should mix args for join and exit

Reply:
I reimplemented the automated join to occur only when a bid is starved for Dai.  This eliminates some configuration complexity.